### ISSUE TITLE
gradient of gradient is still None after second backward

### ISSUE THREAD

**Issue by caiqi** [pytorch/pytorch#28301]
## To Reproduce
```
Steps to reproduce the behavior:
import torch

torch.manual_seed(17)
data = torch.rand(12, 12)
model = torch.nn.Linear(12, 1)
out = model(data)
loss = torch.sum(out ** 2)
grad = torch.autograd.grad(loss, list(model.parameters()), retain_graph=True, create_graph=True)
grad_norm = sum(grad).sum()
grad_norm.backward()

print([m.grad for m in model.parameters()])
print([m.grad for m in grad])
```

```
[tensor([[222.4856, 167.0662, 230.2534, 154.1671, 130.3003, 192.6358, 136.5323,
         169.8731, 174.1054, 171.5416, 220.1357, 173.0526]]), tensor([411.2426])]
[None, None]
```

To get the gradient of the gradient, I tried to use the norm of gradient as loss and backward the gradient norm. However, the gradient of the gradient is still None. It seems that the backward works but the grad is not saved.

## Environment

PyTorch version: 1.3.0
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.14.0

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.1.168
GPU models and configuration: GPU 0: GeForce GTX 1070
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.2

Versions of relevant libraries:
[pip3] numpy==1.16.3
[pip3] numpydoc==0.8.0
[pip3] torch==1.3.0
[pip3] torch-cluster==1.3.0
[pip3] torch-geometric==1.2.1
[pip3] torch-nightly==1.1.0.dev20190417
[pip3] torch-scatter==1.2.0
[pip3] torch-sparse==0.4.0
[pip3] torch-spline-conv==1.1.0
[pip3] torchfile==0.1.0
[pip3] torchnet==0.0.4
[pip3] torchvision==0.4.0
[conda] _tflow_select             2.3.0                       mkl    https://mirrors.ustc.edu.cn/anaconda/pkgs/main
[conda] blas                      1.0                         mkl    https://mirrors.ustc.edu.cn/anaconda/pkgs/main
[conda] mkl                       2019.3                      199    https://mirrors.ustc.edu.cn/anaconda/pkgs/main
[conda] mkl-include               2019.3                      199    https://mirrors.ustc.edu.cn/anaconda/pkgs/main
[conda] mkl-service               1.1.2            py37he904b0f_5    https://mirrors.ustc.edu.cn/anaconda/pkgs/main
[conda] mkl_fft                   1.0.10           py37ha843d7b_0    https://mirrors.ustc.edu.cn/anaconda/pkgs/main
[conda] mkl_random                1.0.2            py37hd81dba3_0    https://mirrors.ustc.edu.cn/anaconda/pkgs/main
[conda] torch                     1.3.0                    pypi_0    pypi
[conda] torch-cluster             1.3.0                    pypi_0    pypi
[conda] torch-geometric           1.2.1                    pypi_0    pypi
[conda] torch-nightly             1.1.0.dev20190417          pypi_0    pypi
[conda] torch-scatter             1.2.0                    pypi_0    pypi
[conda] torch-sparse              0.4.0                    pypi_0    pypi
[conda] torch-spline-conv         1.1.0                    pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchnet                  0.0.4                    pypi_0    pypi
[conda] torchvision               0.4.0                    pypi_0    pypi

**Comment by SsnL** [comment]
Here your `grad` are just intermediate results, not leafs. Gradients by default are not accumulated for leaf tensors. If you need them, use `retain_grad`. Although most of the times you should just use `torch.autograd.grad` again.
